src/helper/class-ee-site.php

```diff
@@ -868,6 +871,173 @@ public function ssl( $args = [], $assoc_args = [], $www_or_non_www = false ) {
+			} else {
+				EE::error( $this->site_data->site_url . ' does not have active publish running.' );
+			}
+
+			if ( ! $refresh ) {
+				return;
+			}
+		}
+
+		if ( $this->site_data->site_ssl ) {
+			EE::error( 'site publish is not yet supported for ssl sites.' );
+		}
+
+		$this->maybe_setup_ngrok( $ngrok );
+		if ( ! empty( $active_publish ) ) {
+			if ( ( $this->site_data->site_url === $active_publish ) ) {
+				$error = $refresh ? '' : "{$this->site_data->site_url} has already been published. Visit link: $publish_url to view it online.\nNote: This link is only valid for few hours. In case it has expired run: `ee site publish {$this->site_data->site_url} --refresh`";
+			} else {
+				$error = "$active_publish site is published currently. Publishing of only one site at a time is supported.\nTo publish {$this->site_data->site_url} , first run: `ee site publish $active_publish --disable`";
+			}
+			if ( ! empty( $error ) ) {
+				EE::error( $error );
+			}
+		}
+
+		if ( ! empty( $token ) ) {
+			EE::exec( "$ngrok authtoken $token" );
+		}
+		$config_80_port = \EE\Utils\get_config_value( 'proxy_80_port', 80 );
+		if ( ! $refresh ) {
+			EE::log( "Publishing site: {$this->site_data->site_url} online." );
+		}
+		EE::debug( "$ngrok http -host-header={$this->site_data->site_url} $config_80_port > /dev/null &" );
+		EE::debug( shell_exec( "$ngrok http -host-header={$this->site_data->site_url} $config_80_port > /dev/null &" ) );
+		$published_url = $this->ngrok_curl();
+		if ( empty( $published_url ) ) {
+			EE::error( 'Could not publish site.' );
+		}
+		EE::success( "Successfully published {$this->site_data->site_url} to url: $published_url" );
+		Option::set( 'publish_site', $this->site_data->site_url );
+		Option::set( 'publish_url', $published_url );
+	}
+
+	/**
+	 * Check if ngrok binary is setup. If not, set it up.
+	 *
+	 * @param $ngrok Path to ngrok binary.
+	 */
+	private function maybe_setup_ngrok( $ngrok ) {
+
+		if ( $this->fs->exists( $ngrok ) ) {
+			return;
+		}
+		EE::log( 'Setting up ngrok. This may take some time.' );
+		$ngrok_zip = $ngrok . '.zip';
+		$this->fs->mkdir( dirname( $ngrok ) );
+		$uname_m      = 'x86_64' === php_uname( 'm' ) ? 'amd64.zip' : '386.zip';
+		$download_url = IS_DARWIN ? 'https://bin.equinox.io/c/4VmDzA7iaHb/ngrok-stable-darwin-' . $uname_m : 'https://bin.equinox.io/c/4VmDzA7iaHb/ngrok-stable-linux-' . $uname_m;
+		download( $ngrok_zip, $download_url );
+		extract_zip( $ngrok_zip, dirname( $ngrok ) );
+		chmod( $ngrok, 0755 );
+		$this->fs->remove( $ngrok_zip );
+	}
+
+	/**
+	 * Function to curl and get data from ngrok api.
+	 *
+	 * @param bool $get_url To get url of tunnel or not.
+	 * @param bool $refresh Whether to disable for publish refresh or not.
+	 */
+	private function ngrok_curl( $get_url = true, $refresh = false ) {
+
+		EE::log( 'Checking ngrok api for tunnel details.' );
+		$tries = 0;
+		$loop  = true;
+		while ( $loop ) {
+			$ch = curl_init( 'http://127.0.0.1:4040/api/tunnels/' );
```

Comment:
This is the local api of ngrok running on host machine. It will work regardless of site being http or https. JFYI, publish feature is [not available for ssl site](https://github.com/EasyEngine/site-command/pull/260/files#diff-39826514bd782321485225d4c4daec7bR930) anyways.